src/tests/baseservices/compilerservices/UnsafeAccessors/UnsafeAccessorsTests.Generics.cs

```diff
@@ -0,0 +1,466 @@
+            Assert.True(Accessors<Struct>.CanCastToElementType<Struct>(a, new Struct()));
+            Assert.Equal(0, a.Count);
+            Accessors<Struct>.Add(a, new Struct());
+            Accessors<Struct>.AddWithIgnore<int>(a, new Struct(), 1);
+            Accessors<Struct>.AddWithIgnore<string>(a, new Struct(), string.Empty);
+            Accessors<Struct>.AddWithIgnore<Struct>(a, new Struct(), new Struct());
+            Assert.Equal(4, a.Count);
+        }
+    }
+
+    [Fact]
+    [ActiveIssue("https://github.com/dotnet/runtime/issues/89439", TestRuntimes.Mono)]
+    public static void Verify_Generic_GenericTypeNonGenericStaticMethod()
+    {
+        Console.WriteLine($"Running {nameof(Verify_Generic_GenericTypeNonGenericStaticMethod)}");
+        {
+            Assert.Equal(typeof(int), Accessors<int>.ElementType(null));
+            Assert.Equal(typeof(string), Accessors<string>.ElementType(null));
+            Assert.Equal(typeof(Struct), Accessors<Struct>.ElementType(null));
+        }
+    }
+
+    [Fact]
+    [ActiveIssue("https://github.com/dotnet/runtime/issues/89439", TestRuntimes.Mono)]
+    public static void Verify_Generic_GenericTypeGenericStaticMethod()
+    {
+        Console.WriteLine($"Running {nameof(Verify_Generic_GenericTypeGenericStaticMethod)}");
+        {
+            Assert.True(Accessors<int>.CanUseElementType<int>(null, 1));
+            Assert.False(Accessors<int>.CanUseElementType<string>(null, string.Empty));
+            Assert.False(Accessors<int>.CanUseElementType<Struct>(null, new Struct()));
+        }
+        {
+            Assert.False(Accessors<string>.CanUseElementType<int>(null, 1));
+            Assert.True(Accessors<string>.CanUseElementType<string>(null, string.Empty));
+            Assert.False(Accessors<string>.CanUseElementType<Struct>(null, new Struct()));
+        }
+        {
+            Assert.False(Accessors<Struct>.CanUseElementType<int>(null, 1));
+            Assert.False(Accessors<Struct>.CanUseElementType<string>(null, string.Empty));
+            Assert.True(Accessors<Struct>.CanUseElementType<Struct>(null, new Struct()));
+        }
+    }
+
+    class ClassWithConstraints
+    {
+        private string M<T, U>() where T : U, IEquatable<T>
+            => $"{typeof(T)}|{typeof(U)}";
+
+        private static string SM<T, U>() where T : U, IEquatable<T>
+            => $"{typeof(T)}|{typeof(U)}";
+    }
+
+    [Fact]
+    [ActiveIssue("https://github.com/dotnet/runtime/issues/89439", TestRuntimes.Mono)]
+    public static void Verify_Generic_ConstraintEnforcement()
+    {
+        Console.WriteLine($"Running {nameof(Verify_Generic_ConstraintEnforcement)}");
+
+        Assert.Equal($"{typeof(string)}|{typeof(object)}", CallMethod<string, object>(new ClassWithConstraints()));
+        Assert.Equal($"{typeof(string)}|{typeof(object)}", CallStaticMethod<string, object>(null));
+
+        // Constraint validation isn't performed in AOT scenarios.
```

Comment:
Is this a bug or expected behavior?

We do not try to be robust against manually written invalid IL in AOT compilers. However, we try to be robust against what you can write in plain C#. For example, reflection in native AOT does constraint validation and this feature is equivalent to reflection.

Reply:
It is as unsafe as reflection where we do constraint validation. There are multiple levels of unsafe.

Reply:
It sounds like you aren't convinced this is acceptable and would prefer me to update native AOT. Is that fair?

Reply:
Yes, I am not convinced by the reasoning behind this choice. I would prefer that we do the constraint validation in the AOT compiler too.

Reply:
Done.